File: llmem/url_validate.py
"""URL validation for urllib calls — prevent SSRF and file-scheme attacks."""

import http.client
import ipaddress
import socket
import urllib.request
import urllib.error
from urllib.parse import urlparse, urlunparse, unquote

OLLAMA_DEFAULT_PORT = 11434

# Default timeout for all urllib calls (seconds).
# Prevents infinite hangs on unresponsive hosts.
DEFAULT_URLOPEN_TIMEOUT = 30


def _ip_is_blocked(ip: ipaddress.IPv4Address | ipaddress.IPv6Address) -> bool:
    """Return True if the IP address should be blocked (private/link-local/etc).

    Loopback addresses are NOT blocked — Ollama typically runs on localhost.
    """
    if ip.is_loopback:
        return False
    return bool(ip.is_private or ip.is_link_local or ip.is_reserved or ip.is_multicast)


def _resolve_hostname(hostname: str) -> list[str] | None:
    """Resolve a hostname to a list of IP address strings.

    Returns None on resolution failure.
    """
    try:
        addrs = socket.getaddrinfo(hostname, None, socket.AF_UNSPEC, socket.SOCK_STREAM)
        return [a[4][0] for a in addrs]
    except Exception:
        return None


def _check_ip_access(
    ip: ipaddress.IPv4Address | ipaddress.IPv6Address,
    allow_remote: bool,
    port: int,
) -> bool:
    """Check whether a resolved IP is permitted for access.

    When allow_remote is False (the default), only loopback addresses on the
    Ollama default port are permitted. All other IPs — including public IPs
    that are not blocked by _ip_is_blocked — are rejected.

    When allow_remote is True, any IP that is not blocked (private, link-local,
    reserved, multicast) is permitted.

    Returns True if allowed, False if blocked.

    Loopback addresses are only permitted on the Ollama default port,
    regardless of allow_remote. The allow_remote flag controls whether
    non-loopback public addresses are allowed — it does NOT open up
    all loopback ports.
    """
    if _ip_is_blocked(ip):
        return False
    if ip.is_loopback:
        if port != OLLAMA_DEFAULT_PORT:
            return False
    elif not allow_remote:
        # Non-loopback addresses require allow_remote=True
        return False
    return True


def _get_effective_port(parsed) -> int:
    """Return the port for a parsed URL, using scheme defaults if missing."""
    if parsed.port is not None:
        return parsed.port
    if parsed.scheme == "https":
        return 443
    return 80


def _strip_credentials(url: str) -> str:
    """Remove userinfo (credentials) from a URL for safe error display.

    Turns 'http://user:pass@host/path' into 'http://host/path'.
    """
    parsed = urlparse(url)
    # Rebuild URL without userinfo
    safe = urlunparse(
        (
            parsed.scheme,
            parsed.hostname or "",
            parsed.path,
            parsed.params,
            parsed.query,
            parsed.fragment,
        )
    )
    port_str = f":{parsed.port}" if parsed.port else ""
    # Reconstruct with hostname + port but no credentials
    netloc = f"{parsed.hostname or ''}{port_str}"
    return urlunparse(
        (
            parsed.scheme,
            netloc,
            parsed.path,
            parsed.params,
            parsed.query,
            parsed.fragment,
        )
    )


def is_safe_url(url: str, allow_remote: bool = False) -> bool:
    """Validate that a URL is safe to pass to urllib.

    Checks:
    - Scheme must be http or https (blocks file://, ftp://, data://, etc.)
    - Must have a hostname
    - Percent-encoded hostnames are decoded before IP checks to prevent
      SSRF bypass (e.g. %31%32%37%2e%30%2e%30%2e%31 → 127.0.0.1).
    - If allow_remote is False (default), only loopback addresses on the
      Ollama default port are allowed.
    - If allow_remote is True, any reachable hostname is allowed (still
      blocks non-http schemes and obviously invalid hostnames).
    - Validates resolved IP to prevent DNS-rebinding TOCTOU: the hostname
      must resolve to permitted addresses immediately before the request.
    """
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https"):
        return False
    hostname = parsed.hostname
    if not hostname:
        return False
    # Reject URLs with credentials embedded — these should never reach urllib
    if parsed.username or parsed.password:
        return False
    port = _get_effective_port(parsed)
    # Percent-decode the hostname before IP checks to prevent SSRF bypass
    # via encoded IP addresses (e.g. %31%32%37%2e%30%2e%30%2e%31 → 127.0.0.1).
    # urllib normalizes percent-encoded hostnames, so we must check the
    # decoded form to match what urllib will actually connect to.
    decoded_hostname = unquote(hostname)
    try:
        ip = ipaddress.ip_address(decoded_hostname)
        return _check_ip_access(ip, allow_remote, port)
    except ValueError:
        resolved = _resolve_hostname(decoded_hostname)
        if resolved is None:
            if not allow_remote:
                return False
            # If DNS fails but the decoded hostname looks like a raw IP,
            # block it — DNS failure on a percent-encoded IP means urllib
            # may still connect to the decoded private IP.
            try:
                ip = ipaddress.ip_address(decoded_hostname)
                return _check_ip_access(ip, allow_remote, port)
            except ValueError:
                pass
        else:
            for addr in resolved:
                try:
                    ip = ipaddress.ip_address(addr)
                except ValueError:
                    return False
                if not _check_ip_access(ip, allow_remote, port):
                    return False
    return True


class _NoRedirectHandler(urllib.request.HTTPRedirectHandler):
    """Redirect handler that blocks all HTTP redirects.

    Prevents SSRF via redirect: an attacker-controlled URL may redirect
    to an internal service (e.g. http://169.254.169.254/latest/meta-data/).
    By blocking all redirects, we ensure the URL resolved at validation
    time is the same one actually fetched.
    """

    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None  # Block all redirects


def _extract_url_string(url: str | urllib.request.Request) -> str:
    """Extract the URL string from a Request object or return the string as-is.

    Callers may pass either a URL string or a urllib.request.Request object
    to safe_urlopen(). This helper normalises the input so that downstream
    validation functions (which expect str) always receive a string.

    Args:
        url: A URL string or a urllib.request.Request object.

    Returns:
        The URL string.

    Raises:
        ValueError: If url is a Request with no full_url attribute.
    """
    if isinstance(url, urllib.request.Request):
        return url.full_url
    return url


def safe_urlopen(
    url: str | urllib.request.Request,
    timeout: int = DEFAULT_URLOPEN_TIMEOUT,
    allow_remote: bool = False,
    **kwargs,
) -> http.client.HTTPResponse:
    """Open a URL with SSRF protections: validates URL, blocks redirects, sets timeout.

    This is the safe replacement for urllib.request.urlopen(). It:
    1. Validates the URL via is_safe_url()
    2. Builds an opener that blocks HTTP redirects (prevents redirect-based SSRF)
    3. Re-resolves the hostname immediately before the request (mitigates DNS rebinding TOCTOU)
    4. Enforces a timeout (prevents infinite hangs)
    5. Strips credentials from error messages

    Args:
        url: The URL to open — either a string or a urllib.request.Request
            object. Must pass is_safe_url() validation.
        timeout: Request timeout in seconds. Defaults to 30.
        allow_remote: If True, allow non-loopback addresses. Must match the
            policy used during URL construction (e.g. ExtractionEngine passes
            allow_remote=True because Ollama can run on remote hosts). Defaults
            to False for safety.
        **kwargs: Additional arguments passed to the opener's open() method.

    Returns:
        An http.client.HTTPResponse object (context manager).

    Raises:
        ValueError: If the URL fails is_safe_url() validation.
        urllib.error.HTTPError: If the server returns an HTTP error (credentials stripped).
        urllib.error.URLError: If the connection fails (credentials stripped).
    """
    url_str = _extract_url_string(url)

    if not is_safe_url(url_str, allow_remote=allow_remote):
        raise ValueError(
            f"llmem: url_validate: URL rejected: must be http(s) to a permitted "
            f"address — got {_strip_credentials(url_str)!r}"
        )

    # Re-resolve the hostname immediately before the request to mitigate
    # DNS rebinding TOCTOU: an attacker might change DNS after validation.
    # Structured to match is_safe_url()'s pattern: ValueError from
    # ipaddress.ip_address() is caught separately so it does not swallow
    # the deliberate ValueError raised for blocked re-resolved addresses.
    parsed = urlparse(url_str)
    hostname = parsed.hostname
    if hostname:
        # Use percent-decoded hostname for re-resolution (matches is_safe_url behavior)
        decoded_hostname = unquote(hostname)
        port = _get_effective_port(parsed)
        resolved = _resolve_hostname(decoded_hostname)
        if resolved:
            for addr in resolved:
                try:
                    ip = ipaddress.ip_address(addr)
                except ValueError:
                    # Unparseable address — treat as blocked.
                    raise ValueError(
                        f"llmem: url_validate: URL rejected after re-resolve: "
                        f"hostname {hostname!r} resolved to unparseable address "
                        f"{addr!r} — got {_strip_credentials(url_str)!r}"
                    )
                if not _check_ip_access(ip, allow_remote, port):
                    raise ValueError(
                        f"llmem: url_validate: URL rejected after re-resolve: "
                        f"hostname {hostname!r} resolved to blocked address {addr} "
                        f"— got {_strip_credentials(url_str)!r}"
                    )

    no_redirect_opener = urllib.request.build_opener(_NoRedirectHandler)

    try:
        return no_redirect_opener.open(url, timeout=timeout, **kwargs)
    except urllib.error.HTTPError as e:
        raise urllib.error.HTTPError(
            _strip_credentials(url_str), e.code, e.reason, e.headers, e.fp
        ) from e
    except urllib.error.URLError as e:
        safe_url = _strip_credentials(url_str)
        raise urllib.error.URLError(
            f"request to {safe_url!r} failed: {e.reason}"
        ) from e


def validate_url(url: str, allow_remote: bool = False) -> str:
    """Validate URL and return it, or raise ValueError."""
    if not is_safe_url(url, allow_remote=allow_remote):
        raise ValueError(
            f"URL rejected: must be http(s) to a permitted address — got {_strip_credentials(url)!r}"
        )
    return url


def validate_base_url(base_url: str, module: str = "url_validate") -> str:
    """Validate and normalize an Ollama base URL.

    Performs three checks shared by EmbeddingEngine, ExtractionEngine,
    and IntrospectionAnalyzer constructors:
    1. Strips trailing slash
    2. Validates http:// or https:// prefix
    3. Validates via is_safe_url with allow_remote=True

    Args:
        base_url: The Ollama base URL to validate.
        module: Module name for error messages (e.g. 'embed', 'extract', 'introspection').

    Returns:
        The stripped, validated URL string.

    Raises:
        ValueError: If the URL is not http(s) or fails is_safe_url().
    """
    base_url = base_url.rstrip("/")
    if not base_url.startswith(("http://", "https://")):
        raise ValueError(
            f"llmem: {module}: unsafe Ollama URL (must be http/https): {_strip_credentials(base_url)!r}"
        )
    if not is_safe_url(base_url, allow_remote=True):
        raise ValueError(
            f"llmem: {module}: unsafe Ollama URL (blocked address): {_strip_credentials(base_url)!r}"
        )
    return base_url

File: internal/urlvalidate/urlvalidate.go
// Package urlvalidate provides SSRF-protected URL validation and safe HTTP access.
// It blocks private/link-local IPs, percent-encoded SSRF bypasses, and redirect-based
// attacks, while allowing loopback on the Ollama default port for local development.
package urlvalidate

import (
	"context"
	"fmt"
	"log/slog"
	"net"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// DefaultOllamaPort is the default port for Ollama API.
const DefaultOllamaPort = 11434

// DefaultURLTimeout is the default timeout for HTTP requests.
const DefaultURLTimeout = 30 * time.Second

// fmtErr wraps an error with the "llmem: urlvalidate:" domain prefix.
func fmtErr(format string, args ...any) error {
	return fmt.Errorf("llmem: urlvalidate: "+format, args...)
}

// isPrivateIP checks if an IP is a private, link-local, reserved, or multicast address.
// Loopback addresses return false — they are handled separately.
func isPrivateIP(ip net.IP) bool {
	if ip.IsLoopback() {
		return false
	}
	return ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() ||
		ip.IsUnspecified() || ip.IsMulticast()
}

// checkIPAccess checks whether a resolved IP is permitted for access.
//
// When allowRemote is false (the default), only loopback addresses on the
// Ollama default port are permitted. All other IPs — including public IPs
// that are not blocked by isPrivateIP — are rejected.
//
// When allowRemote is true, any IP that is not private/link-local/reserved/multicast
// is permitted. Loopback addresses are still restricted to the Ollama default port
// regardless of allowRemote.
func checkIPAccess(ip net.IP, allowRemote bool, port int) bool {
	if isPrivateIP(ip) {
		return false
	}
	if ip.IsLoopback() {
		if port != DefaultOllamaPort {
			return false
		}
	} else if !allowRemote {
		return false
	}
	return true
}

// resolveHostname resolves a hostname to a list of IP addresses.
// Returns nil if resolution fails.
func resolveHostname(hostname string) []net.IP {
	addrs, err := net.LookupIP(hostname)
	if err != nil {
		return nil
	}
	return addrs
}

// getEffectivePort returns the port for a parsed URL, using scheme defaults if missing.
func getEffectivePort(u *url.URL) int {
	if u.Port() != "" {
		port, err := net.LookupPort("tcp", u.Port())
		if err == nil {
			return port
		}
	}
	if u.Scheme == "https" {
		return 443
	}
	return 80
}

// IsSafeURL validates that a URL is safe to access.
//
// It checks: scheme is http or https, hostname is present, no embedded credentials,
// resolved IPs are not blocked. Loopback is allowed on the Ollama default port
// when allowRemote is false. Percent-encoded hostnames are decoded before IP checks
// to prevent SSRF bypass.
//
// Never panics. Returns false for malformed URLs.
func IsSafeURL(urlStr string, allowRemote bool) bool {
	u, err := url.Parse(urlStr)
	if err != nil {
		return false
	}

	if u.Scheme != "http" && u.Scheme != "https" {
		return false
	}

	hostname := u.Hostname()
	if hostname == "" {
		return false
	}

	// Reject URLs with embedded credentials
	if u.User != nil {
		return false
	}

	port := getEffectivePort(u)

	// Percent-decode the hostname before IP checks to prevent SSRF bypass
	// via encoded IP addresses (e.g. %31%32%37%2e%30%2e%30%2e%31 → 127.0.0.1).
	decodedHostname := decodeHostname(hostname)

	// Try parsing as an IP address first
	if ip := net.ParseIP(decodedHostname); ip != nil {
		return checkIPAccess(ip, allowRemote, port)
	}

	// Resolve hostname and check all IPs
	addrs := resolveHostname(decodedHostname)
	if addrs == nil {
		if !allowRemote {
			return false
		}
		// DNS failed but allowRemote — check decoded hostname as raw IP again
		if ip := net.ParseIP(decodedHostname); ip != nil {
			return checkIPAccess(ip, allowRemote, port)
		}
		// DNS failed and not an IP — allow if allowRemote (could be a domain)
		return true
	}

	for _, addr := range addrs {
		if !checkIPAccess(addr, allowRemote, port) {
			return false
		}
	}
	return true
}

// decodeHostname percent-decodes a hostname for SSRF bypass prevention.
// Handles percent-encoded octets like %31%32%37 → "127".
func decodeHostname(hostname string) string {
	// Use url.PathUnescape which decodes percent-encoded sequences
	decoded, err := url.PathUnescape(hostname)
	if err != nil {
		// If decoding fails, return the hostname as-is
		return hostname
	}
	return decoded
}

// noRedirectTransport is an unexported http.RoundTripper that blocks redirects
// for SSRF prevention. An attacker-controlled URL may redirect to an internal
// service (e.g. http://169.254.169.254/latest/meta-data/).
type noRedirectTransport struct {
	base http.RoundTripper
}

func (t noRedirectTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	resp, err := t.base.RoundTrip(req)
	if err != nil {
		return nil, err
	}

	// Check if the response is a redirect
	if resp.StatusCode >= 300 && resp.StatusCode < 400 {
		location := resp.Header.Get("Location")
		safeURL := stripCredentials(req.URL.String())
		slog.Warn("llmem: urlvalidate: blocked SSRF redirect", "from", safeURL, "to", location)
		// Close the response body to avoid resource leaks
		resp.Body.Close()
		return nil, fmtErr("blocked redirect from %s to %s", safeURL, location)
	}
	return resp, nil
}

// SafeURLOpen opens a URL with SSRF protections.
//
// It validates the URL via IsSafeURL, blocks redirects, re-resolves the
// hostname before the request (DNS rebinding protection), and enforces a
// timeout. Returns *http.Response on success — the caller MUST close the
// response body.
//
// If the URL fails IsSafeURL, returns an error with the domain prefix.
// Context cancellation is respected.
func SafeURLOpen(ctx context.Context, urlStr string, timeout time.Duration, allowRemote bool) (*http.Response, error) {
	if !IsSafeURL(urlStr, allowRemote) {
		return nil, fmtErr("URL rejected: must be http(s) to a permitted address — got %s", stripCredentials(urlStr))
	}

	if timeout <= 0 {
		timeout = DefaultURLTimeout
	}

	u, err := url.Parse(urlStr)
	if err != nil {
		return nil, fmtErr("parse URL: %w", err)
	}

	// Re-resolve the hostname immediately before the request to mitigate
	// DNS rebinding TOCTOU: an attacker might change DNS after validation.
	hostname := u.Hostname()
	if hostname != "" {
		decodedHostname := decodeHostname(hostname)
		port := getEffectivePort(u)
		addrs := resolveHostname(decodedHostname)
		if addrs != nil {
			for _, addr := range addrs {
				if !checkIPAccess(addr, allowRemote, port) {
					return nil, fmtErr("URL rejected after re-resolve: hostname %q resolved to blocked address %s — got %s", hostname, addr, stripCredentials(urlStr))
				}
			}
		}
	}

	client := &http.Client{
		Transport: noRedirectTransport{base: http.DefaultTransport},
		Timeout:   timeout,
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, urlStr, nil)
	if err != nil {
		return nil, fmtErr("create request: %w", err)
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmtErr("request to %s failed: %w", stripCredentials(urlStr), err)
	}

	return resp, nil
}

// ValidateBaseURL validates and normalizes an Ollama base URL.
//
// It strips the trailing slash, validates the http/https scheme, and validates
// the URL via IsSafeURL with allowRemote=true. Returns the validated URL string.
// Error messages include the module parameter for context.
func ValidateBaseURL(baseURL, module string) (string, error) {
	baseURL = strings.TrimRight(baseURL, "/")
	if !strings.HasPrefix(baseURL, "http://") && !strings.HasPrefix(baseURL, "https://") {
		return "", fmtErr("%s: unsafe Ollama URL (must be http/https): %s", module, stripCredentials(baseURL))
	}
	// Validate with allowRemote=true since Ollama can run on remote hosts
	if !IsSafeURL(baseURL, true) {
		return "", fmtErr("%s: unsafe Ollama URL (blocked address): %s", module, stripCredentials(baseURL))
	}
	return baseURL, nil
}

// IsRemoteAllowed infers allowRemote from a URL.
// Returns false for loopback addresses, private/link-local IPs, and local hostnames.
// Returns true for public IPs and hostnames that resolve to public IPs.
// Returns false (fail-closed) for hostnames that fail DNS resolution.
func IsRemoteAllowed(urlStr string) bool {
	u, err := url.Parse(urlStr)
	if err != nil {
		return false
	}
	hostname := u.Hostname()
	if hostname == "" {
		return false
	}

	// Check if it's an IP address
	if ip := net.ParseIP(hostname); ip != nil {
		return !ip.IsLoopback() && !isPrivateIP(ip)
	}

	// Check for local hostnames
	localHostnames := map[string]bool{
		"localhost":             true,
		"localhost.localdomain": true,
		"localhost6":            true,
	}
	if localHostnames[strings.ToLower(hostname)] {
		return false
	}

	// Resolve the hostname and check all resulting IPs.
	// If DNS resolution fails, fail-closed (return false).
	addrs := resolveHostname(hostname)
	if addrs == nil {
		return false
	}
	for _, addr := range addrs {
		if addr.IsLoopback() || isPrivateIP(addr) {
			return false
		}
	}
	return true
}

// stripCredentials removes userinfo (credentials) from a URL for safe error display.
// Turns 'http://user:pass@host/path' into 'http://host/path'.
func stripCredentials(urlStr string) string {
	u, err := url.Parse(urlStr)
	if err != nil {
		return urlStr // Return as-is if parsing fails
	}
	// Reconstruct with hostname + port but no credentials
	return u.Scheme + "://" + u.Host + u.Path
	// Note: u.Host already excludes userinfo
}